In [3]:
## Assignment 9
## Power Iteration Method
import numpy as np

In [4]:
## start vector x and a is transition matrix which is given in th question
x = np.array([0.1666,0.1666,0.1666,0.1666,0.1666,0.1666])
a = np.array([[0.0333,0.0333,0.4333,0.0333,0.0333,0.4333],
              [0.0333,0.0333,0.0333,0.0333,0.0333,0.8333],
              [0.0333,0.8333,0.0333,0.0333,0.0333,0.0333],
              [0.4333,0.4333,0.0333,0.0333,0.0333,0.0333],
              [0.0333,0.0333,0.3,0.3,0.0333,0.3],
              [0.1666,0.1666,0.1666,0.1666,0.1666,0.1666]
            
            ])

## .T means transpose


In [5]:
a*x

array([[0.00554778, 0.00554778, 0.07218778, 0.00554778, 0.00554778,
        0.07218778],
       [0.00554778, 0.00554778, 0.00554778, 0.00554778, 0.00554778,
        0.13882778],
       [0.00554778, 0.13882778, 0.00554778, 0.00554778, 0.00554778,
        0.00554778],
       [0.07218778, 0.07218778, 0.00554778, 0.00554778, 0.00554778,
        0.00554778],
       [0.00554778, 0.00554778, 0.04998   , 0.04998   , 0.00554778,
        0.04998   ],
       [0.02775556, 0.02775556, 0.02775556, 0.02775556, 0.02775556,
        0.02775556]])

In [6]:
A = np.matrix(a)  ##converting transpose matrix vakues into float values matrix
X0 = np.matrix([1/6,1/6,1/6,1/6,1/6,1/6]).T
iteration = 5     ## depend on the number of iteration required in question

In [7]:
A

matrix([[0.0333, 0.0333, 0.4333, 0.0333, 0.0333, 0.4333],
        [0.0333, 0.0333, 0.0333, 0.0333, 0.0333, 0.8333],
        [0.0333, 0.8333, 0.0333, 0.0333, 0.0333, 0.0333],
        [0.4333, 0.4333, 0.0333, 0.0333, 0.0333, 0.0333],
        [0.0333, 0.0333, 0.3   , 0.3   , 0.0333, 0.3   ],
        [0.1666, 0.1666, 0.1666, 0.1666, 0.1666, 0.1666]])

In [9]:
# ax where x defines the number of iterations
a1 = X0.T*A
a2 = X0.T*A**2
a3 = X0.T*A**3
a4 = X0.T*A**4
a5 = X0.T*A**5

a10 = X0.T*A**10
a11 = X0.T*A**11
a12 = X0.T*A**12

In [10]:
print(a1)
print(a2)
print(a3)
print(a4)
print(a5)
 
## printing out the desired values for here it was 5 iteration "a5" is the answere for it


[[0.12218333 0.25551667 0.16663333 0.09996667 0.05551667 0.29996667]]
[[0.11326501 0.24657168 0.13695797 0.08808464 0.07327834 0.3413713 ]]
[[0.11402296 0.22358934 0.14363845 0.09833244 0.07878911 0.34089579]]
[[0.11805001 0.23296077 0.14533928 0.09973009 0.07871704 0.32421075]]
[[0.11637629 0.23264771 0.14469809 0.09747809 0.07648426 0.33106671]]


In [13]:
## For Convergence check the iteration as well as difference between two a's
print("2 iteration",a2-a1,a2-a1 <= [0.001,0.001,0.001,0.001,0.001,0.001])
print("3 iteration",a3-a2,a3-a2 <= [0.001,0.001,0.001,0.001,0.001,0.001])
print("4 iteration",a4-a3,a4-a3 <= [0.001,0.001,0.001,0.001,0.001,0.001])
print("5 iteration",a5-a4,a5-a4 <= [0.001,0.001,0.001,0.001,0.001,0.001])


print("11 iteration",a11-a10,a11-a10 <= [0.001,0.001,0.001,0.001,0.001,0.001])

print("12 iteration",a12-a11,a12-a11 <= [0.001,0.001,0.001,0.001,0.001,0.001])

2 iteration [[-0.00891833 -0.00894499 -0.02967536 -0.01188203  0.01776167  0.04140464]] [[ True  True  True  True False False]]
3 iteration [[ 0.00075795 -0.02298234  0.00668048  0.01024781  0.00551077 -0.00047552]] [[ True  True False False False  True]]
4 iteration [[ 4.02704920e-03  9.37142946e-03  1.70083037e-03  1.39764852e-03
  -7.20729291e-05 -1.66850385e-02]] [[False False False False  True  True]]
5 iteration [[-0.00167372 -0.00031306 -0.00064118 -0.002252   -0.00223278  0.00685596]] [[ True  True  True  True  True False]]
11 iteration [[-3.01076365e-05 -6.11833642e-05 -3.41319460e-05 -2.55847976e-05
  -2.56436157e-05 -8.10965247e-05]] [[ True  True  True  True  True  True]]
12 iteration [[-2.96270903e-05 -5.69326471e-05 -3.82753782e-05 -2.62323236e-05
  -1.93931713e-05 -8.72220695e-05]] [[ True  True  True  True  True  True]]


In [38]:
## Rank Questions of Assignement 9

In [39]:
A = np.array([[1,0,1,0,1],[0,0,0,1,1],[1,0,0,0,1],[0,0,0,1,1],[1,0,0,0,0]])  ##according to question
A = A.astype('float64')

In [40]:
A

array([[1., 0., 1., 0., 1.],
       [0., 0., 0., 1., 1.],
       [1., 0., 0., 0., 1.],
       [0., 0., 0., 1., 1.],
       [1., 0., 0., 0., 0.]])

In [41]:
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np


In [42]:
# ## for drawing the matrix
# G = nx.from_numpy_matrix(np.matrix(A),create_using=nx.DiGraph)
# layout = nx.spring_layout(G)
# nx.draw_networkx_edge_lables(G,pos=layout)
# plt.show()

In [43]:
## converting to adj matrix with transpose
adj_mat =A.T

In [44]:
## Degree Calculation
def deg_calc(adj_mat):
    return(adj_mat.sum(axis=0),adj_mat.sum(axis=1))

In [45]:
in_d,out_d = deg_calc(A)

In [46]:
in_d,out_d

(array([3., 0., 1., 2., 4.]), array([3., 2., 2., 2., 1.]))

In [47]:
in_d == out_d

array([ True, False, False,  True, False])

In [48]:
diff_array = in_d - out_d

In [49]:
##difference between In and Out-Degree
np.absolute(diff_array)

## Range is found from max - min degree i.e 3-0 = 3

array([0., 2., 1., 0., 3.])

In [50]:
### STM  Stochastic Transition Matrix (STM) 
def stm_calc(adj_mat, alpha):
    count = adj_mat.sum(axis = 1)
    print("count of degrees",count)
    z_row = np.where(~adj_mat.any(axis=1))[0]
    print(z_row)
    if(z_row.size):
        for item in z-row:
            #print item
            adj_mat[item,:] = float(1/len(adj_mat))
    print("\n when encouraging a row with all zeroes\n",adj_mat)
    
    for i in range(len(count)):
        if(i in z_row):
            continue
        else:
            adj_mat[i] = adj_mat[i]/count[i]
    
    adj_mat = np.where(np.isnan(adj_mat),0,adj_mat)
    print("\n divide all by the number of ",adj_mat)
    
    adj_mat = (adj_mat*(1-alpha))
    print("\n multiply matrix by 1-alpha of ",adj_mat)
    
    adj_mat = adj_mat + (alpha/len(count))
    print("\n adding alpha/n to all matrix ",adj_mat)
        
    return(adj_mat)    

In [51]:
S_T_M = stm_calc(A,0.1)
print("\n STM \n",S_T_M)

count of degrees [3. 2. 2. 2. 1.]
[]

 when encouraging a row with all zeroes
 [[1. 0. 1. 0. 1.]
 [0. 0. 0. 1. 1.]
 [1. 0. 0. 0. 1.]
 [0. 0. 0. 1. 1.]
 [1. 0. 0. 0. 0.]]

 divide all by the number of  [[0.33333333 0.         0.33333333 0.         0.33333333]
 [0.         0.         0.         0.5        0.5       ]
 [0.5        0.         0.         0.         0.5       ]
 [0.         0.         0.         0.5        0.5       ]
 [1.         0.         0.         0.         0.        ]]

 multiply matrix by 1-alpha of  [[0.3  0.   0.3  0.   0.3 ]
 [0.   0.   0.   0.45 0.45]
 [0.45 0.   0.   0.   0.45]
 [0.   0.   0.   0.45 0.45]
 [0.9  0.   0.   0.   0.  ]]

 adding alpha/n to all matrix  [[0.32 0.02 0.32 0.02 0.32]
 [0.02 0.02 0.02 0.47 0.47]
 [0.47 0.02 0.02 0.02 0.47]
 [0.02 0.02 0.02 0.47 0.47]
 [0.92 0.02 0.02 0.02 0.02]]

 STM 
 [[0.32 0.02 0.32 0.02 0.32]
 [0.02 0.02 0.02 0.47 0.47]
 [0.47 0.02 0.02 0.02 0.47]
 [0.02 0.02 0.02 0.47 0.47]
 [0.92 0.02 0.02 0.02 0.02]]


In [52]:
# Page Rank Inintial Vector is 1/n where n is number nodes(rows)
1/5

0.2

In [53]:
#[0.2,0.2,0.2,0.2,0.2] is the initial vector then

In [54]:
def rank_calc(stm,X):
    c = len(stm.sum(axis=1))
    prob = 1/c
    vec_x0 = np.full((1,c),prob)
    print("intial vetor",vec_x0)
    while(X!=0):
        vec_x0 = np.dot(vec_x0,stm)
        X = X-1
        print(vec_x0)

In [55]:
rank_calc(S_T_M,5)

intial vetor [[0.2 0.2 0.2 0.2 0.2]]
[[0.35 0.02 0.08 0.2  0.35]]
[[0.476 0.02  0.125 0.119 0.26 ]]
[[0.45305 0.02    0.1628  0.08255 0.2816 ]]
[[0.482615  0.02      0.155915  0.0661475 0.2753225]]
[[0.4827365  0.02       0.1647845  0.05876638 0.27371263]]


# After 5 itertation PAger rank solution

[[0.4827365  0.02       0.1647845  0.05876638 0.27371263]]
is answerer for